### PR TITLE
encryption: support seek for encryption writer

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What problem does this PR solve?

Problem Summary:

Turns out we still need that `Seek` for encryption to work with raft engine.

### What is changed and how it works?

What's Changed:

Unify implementation for both encryption reader and writer. `CrypterCore` now encapsulate all the encryption core algorithm, and it's seek-able. Implementation details are mostly the same, read/write still uses different code paths:

* reader calls `do_crypter_in_place`, writer calls `do_crypter`. Only the latter will return a slice of internal buffer as encryption result.
* reader use `MAX_INPLACE_CRYPTION_SIZE`(1M) as a upper bound for internal buffer, writer impose no such limit

### Check List 

Tests

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/encryption/src/io.rs [hunk 1]
```diff
@@ -17,70 +17,162 @@ use file_system::File;
 use tikv_util::box_err;
 
 const AES_BLOCK_SIZE: usize = 16;
-const MAX_CRYPTER_READER_BUFFER_SIZE: usize = 1024;
+const MAX_INPLACE_CRYPTION_SIZE: usize = 1024 * 1024;
 
-/// Encrypt content as data being read.
-pub struct EncrypterReader<R>(CrypterReader<R>);
+/// Encrypt or decrypt content as data being read.
+pub struct CrypterReader<R> {
```
Comment (Connor1996) [inline components/encryption/src/io.rs:23]: can we still separate `Encrypter` and `Decrypter`?

Comment (tabokie) [reply components/encryption/src/io.rs]: Make sense, it's separated now.

File: components/encryption/src/io.rs [hunk 2]
```diff
@@ -17,73 +17,303 @@ use file_system::File;
...
+        Ok(EncryptWriter(CrypterWriter::new(
+            writer,
+            method,
+            key,
+            Mode::Encrypt,
+            iv,
+        )?))
+    }
+
+    pub fn finalize(self) -> IoResult<W> {
```
Comment (Connor1996) [inline components/encryption/src/io.rs:117]: why need to return the inner writer

Comment (Connor1996) [reply components/encryption/src/io.rs:117]: should we call flush explicitly?

File: components/encryption/src/io.rs [hunk 3]
```diff
@@ -17,73 +17,303 @@ use file_system::File;
 use tikv_util::box_err;
 
 const AES_BLOCK_SIZE: usize = 16;
-const MAX_CRYPTER_READER_BUFFER_SIZE: usize = 1024;
```
Comment (Connor1996) [inline components/encryption/src/io.rs:20]: why remove

File: components/encryption/src/io.rs [hunk 4]
```diff
@@ -106,63 +336,50 @@ pub fn create_aes_ctr_crypter(
...
-    fn reserve_buffer(&mut self, size: usize) {
+    #[inline]
+    fn reset_buffer(&mut self, size: usize) {
         // OCrypter require the output buffer to have block_size extra bytes, or it will panic.
-        let target = std::cmp::min(size, MAX_CRYPTER_READER_BUFFER_SIZE) + self.block_size;
-        if target > self.crypter_buffer.len() {
-            self.crypter_buffer.resize(target, 0);
+        self.buffer.reserve(size + self.block_size);
+        unsafe {
+            self.buffer.set_len(size + self.block_size);
```
Comment (Connor1996) [inline components/encryption/src/io.rs:369]: not sure if it's okay to not fill the padding of 0. /cc @yiwu-arbug

File: components/encryption/src/io.rs (hunk 2 above)
Comment (tabokie) [reply components/encryption/src/io.rs:117]: It's just an idiomatic to deal with chained streams, e.g. https://docs.rs/lz4/1.15.131/lz4/struct.Encoder.html#method.finish or https://doc.rust-lang.org/std/io/struct.BufWriter.html#method.into_inner.
This function usually only guarantees bytes are written, no need to be flushed.

File: components/encryption/src/io.rs (hunk 3 above)
Comment (tabokie) [reply components/encryption/src/io.rs:20]: It's renamed.

File: components/encryption/src/io.rs (hunk 4 above)
Comment (tabokie) [reply components/encryption/src/io.rs:369]: It's just an output buffer, the behavior is stated in document: https://docs.rs/openssl/0.9.1/openssl/symm/struct.Crypter.html#method.update

File: components/encryption/src/io.rs (hunk 3 above)
Comment (Connor1996) [reply components/encryption/src/io.rs:20]: but with different value

Comment (tabokie) [reply components/encryption/src/io.rs:20]: @yiwu-arbug suggested in last PR

File: components/encryption/src/io.rs [hunk 5]
```diff
@@ -422,25 +553,69 @@ mod tests {
...
+        OsRng.fill_bytes(&mut plaintext);
+        let offsets = [1024, 1024 + 1, 10240 - 1];
+        let sizes = [1024, 8000];
+        let written = vec![0; 10240];
+        for method in methods {
+            let key = generate_data_key(method);
+            let writable_text = std::io::Cursor::new(written.clone());
+            let iv = Iv::new_ctr();
+            let encrypter = EncrypterWriter::new(writable_text, method, &key, iv).unwrap();
+            let mut decrypter = DecrypterWriter::new(encrypter, method, &key, iv).unwrap();
```
Comment (hunterlxt) [inline components/encryption/src/io.rs:594]: Maybe we need to prove here that the previous encryption was successful

Comment (tabokie) [reply components/encryption/src/io.rs:594]: You mean encrypted != plaintext?

File: components/encryption/src/io.rs (hunk 3 above)
Comment (hunterlxt) [reply components/encryption/src/io.rs:20]: @tabokie can you post that

**Review by hunterlxt — APPROVED** [review]
LGTM

**Comment by tabokie** [comment]
Pin @Connor1996

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**